reactor-core/src/main/java/reactor/core/publisher/Tracer.java

```diff
@@ -0,0 +1,166 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import sun.misc.JavaLangAccess;
+import sun.misc.SharedSecrets;
+
+/**
+ * Utility class for the call-site extracting.
+ *
+ * @author Simon Baslé
+ * @author Sergei Egorov
+ */
+final class Tracer {
+
+	/**
+	 * If set to true, the creation of FluxOnAssembly will capture the raw stacktrace
+	 * instead of the sanitized version.
+	 */
+	static final boolean full = Boolean.parseBoolean(System.getProperty(
+			"reactor.trace.assembly.fullstacktrace",
+			"false"));
+
+	/**
+	 * Transform the current stack trace into a {@link String} representation,
+	 * each element being prepended with a tabulation and appended with a
+	 * newline.
+	 */
+	static Supplier<Supplier<String>> callSiteSupplierFactory;
+
+	static {
+		String[] strategyClasses = {
+				Tracer.class.getPackage().getName() + ".StackWalkerTracer",
+				Tracer.class.getName() + "$SharedSecretsCallSiteSupplierFactory",
+				Tracer.class.getName() + "$ExceptionCallSiteSupplierFactory",
+		};
+		callSiteSupplierFactory = Stream
+				.of(strategyClasses)
+				.flatMap(className -> {
+					try {
+						Class<?> clazz = Class.forName(className);
+						@SuppressWarnings("unchecked")
+						Supplier<Supplier<String>> function = (Supplier) clazz.getDeclaredConstructor()
+						                             .newInstance();
+						return Stream.of(function);
+					}
+					catch (Throwable e) {
+						return Stream.empty();
+					}
+				})
+				.findFirst()
+				.orElseThrow(() -> new IllegalStateException("Valid strategy not found"));
+	}
+
+	@SuppressWarnings("unused")
+	static class SharedSecretsCallSiteSupplierFactory implements Supplier<Supplier<String>> {
+
+		@Override
+		public Supplier<String> get() {
+			return new TracingException();
+		}
+
+		static class TracingException extends Throwable implements Supplier<String> {
+
+			static final JavaLangAccess javaLangAccess = SharedSecrets.getJavaLangAccess();
+
+			@Override
+			public String get() {
+				int stackTraceDepth = javaLangAccess.getStackTraceDepth(this);
+
+				StackTraceElement previousElement = null;
+				// Skip get()
+				for (int i = 2; i < stackTraceDepth; i++) {
+					StackTraceElement e = javaLangAccess.getStackTraceElement(this, i);
+
+					String className = e.getClassName();
+					if (Traces.isUserCode(className)) {
+						StringBuilder sb = new StringBuilder();
+
+						if (previousElement != null) {
+							sb.append("\t").append(previousElement.toString()).append("\n");
+						}
+						sb.append("\t").append(e.toString()).append("\n");
+						return sb.toString();
+					}
+					else {
+						if (!full && e.getLineNumber() <= 1) {
```

Comment:
can you double check that skipping the first line is still relevant with the GLOBAL_TRACING ordering change? especially if there are supplementary hooks... here IIRC the goal was to skip the first line which would always be the xxxOnAssembly subscriber.

Reply:
since this is source code's lines, this condition should skip native methods and methods without source code info (lambdas bridges, unknown source, etc)

Reply:
got it 👍